### PR TITLE
ipq807x: prpl-haze: enable LED driver on device tree

### PR DESCRIPTION
**Pull request by CheWei-Chien**
Enable prpl-haze LED driver LP5562 on device tree.

You also need to set:
CONFIG_LEDS_LP5562=y
CONFIG_LEDS_LP55XX_COMMON=y
in kernel config

Signed-off-by: CheWei Chien <chewei.chien@wnc.com.tw>

### CODE REVIEW

File: target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts [hunk 1]
```diff
@@ -78,6 +78,13 @@
 			bias-pull-up;
 		};
 	};
+
+	i2c_0_pins: i2c-0-state {
```
Comment (robimarko) [inline target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:82]: This is not I2C0, so rename it accordingly.

File: target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts [hunk 2]
```diff
@@ -303,3 +310,39 @@
...
 	qcom,ath11k-calibration-variant = "prpl-Haze";
 };
+
+&blsp1_i2c3{
+	pinctrl-0 = <&i2c_0_pins>;
+	pinctrl-names = "default";
+	status = "ok";
+
+	led-controller@30 {
+		compatible = "ti,lp5562";
```
Comment (robimarko) [inline target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:320]: How is the driver kmod included on this device?
I dont see it being added to the default packages

File: target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts [hunk 3]
```diff
@@ -303,3 +310,39 @@
...
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		status_red: chan@0 {
+			chan-name = "red";
+			led-cur = /bits/ 8 <0x20>;
+			max-cur = /bits/ 8 <0x60>;
+			color = <LED_COLOR_ID_RED>;
+			reg = <0>;
+		};
```
Comment (robimarko) [inline target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:332]: Add newline after each subnode

**Comment by CheWei-Chien** [comment]
Thank for your review!
I do some fix in new commit.
I think the driver kmod should enable on openwrt/target/linux/generic/config-6.1 if possible.

**Comment by ynezz** [comment]
> I think the driver kmod should enable on openwrt/target/linux/generic/config-6.1 if possible.

Although we've already following prior example in the [ipq40xx target](https://github.com/openwrt/openwrt/blob/main/target/linux/ipq40xx/config-6.1#L244-L246):

```
target/linux/ipq40xx/config-6.1:CONFIG_LEDS_LP5523=y
target/linux/ipq40xx/config-6.1:CONFIG_LEDS_LP5562=y
target/linux/ipq40xx/config-6.1:CONFIG_LEDS_LP55XX_COMMON=y
```

I think, that the correct approach here would be to add new kernel module package `KernelPackage/leds-lp5562` into `package/kernel/linux/modules/leds.mk` and then add it to `DEVICE_PACKAGES` kernel module list in the image definition.

According to the [DTS checklist](https://openwrt.org/submitting-patches#dts_checklist)
> 4. If it's possible try to dedicate some of the LEDs for system status indication in [diag.sh](https://github.com/openwrt/openwrt/blob/master/package/base-files/files/etc/diag.sh) ([example](https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11.dtsi#L7))

It makes me wonder if it would be possible to define such LED aliases as well?

**Comment by robimarko** [comment]
Please squash the commits, no need for those.

And no, it should not be enabled in the generic config, please add the required kmod as default package for this device

File: target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts [hunk 4]
```diff
@@ -303,3 +310,41 @@
...
+	status = "okay";
+
+	led-controller@30 {
+		compatible = "ti,lp5562";
+		reg = <0x30>;
+		clock-mode = /bits/ 8 <2>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		status_red: chan@0 {
```
Comment (ynezz) [inline target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:326]: Nitpick, there seems to be a convention to name the aliases in the following way

```suggestion
		led_system_red: chan@0 {
```

File: target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts [hunk 5]
```diff
@@ -303,3 +310,41 @@
...
+			chan-name = "red";
+			led-cur = /bits/ 8 <0x20>;
+			max-cur = /bits/ 8 <0x60>;
+			color = <LED_COLOR_ID_RED>;
+			reg = <0>;
+		};
+
+		status_green: chan@1 {
```
Comment (ynezz) [inline target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:334]: Nitpick, there seems to be a convention to name the aliases in the following way

```suggestion
		led_system_green: chan@1 {
```

File: target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts [hunk 6]
```diff
@@ -303,3 +310,41 @@
...
+			chan-name = "green";
+			led-cur = /bits/ 8 <0x20>;
+			max-cur = /bits/ 8 <0x60>;
+			color = <LED_COLOR_ID_GREEN>;
+			reg = <1>;
+		};
+
+		status_blue: chan@2 {
```
Comment (ynezz) [inline target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:342]: Nitpick, there seems to be a convention to name the aliases in the following way

```suggestion
		led_system_blue: chan@2 {
```

File: target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts [hunk 7]
```diff
@@ -303,3 +310,41 @@
...
+			chan-name = "blue";
+			led-cur = /bits/ 8 <0x20>;
+			max-cur = /bits/ 8 <0x60>;
+			color = <LED_COLOR_ID_BLUE>;
+			reg = <2>;
+		};
```
Comment (ynezz) [inline target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:348]: I don't have access to the device now, so can't check it myself, but I'm wondering if the 3rd/white LED channel was left intentionally or its a mistake?

Comment (CheWei-Chien) [reply target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:348]: The 3rd/white LED channel is broken at the physical layer, so I didn't set it.

**Comment by CheWei-Chien** [comment]
Thank for yours reviews!
I squash the commits, add new kernel module package for LED driver and included it on default package for HAZE.
If have any mistake please let me know, thank you.

File: target/linux/qualcommax/image/ipq807x.mk [hunk 8]
```diff
@@ -195,7 +195,7 @@ define Device/prpl_haze
 	DEVICE_DTS_CONFIG := config@hk09
 	SOC := ipq8072
 	DEVICE_PACKAGES += ath11k-firmware-qcn9074 ipq-wifi-prpl_haze kmod-ath11k-pci \
-		mkf2fs f2fsck kmod-fs-f2fs
+		mkf2fs f2fsck kmod-fs-f2fs kmod-leds-lp55xx-common kmod-leds-lp5562
```
Comment (ynezz) [inline target/linux/qualcommax/image/ipq807x.mk:198]:
```suggestion
		mkf2fs f2fsck kmod-fs-f2fs kmod-leds-lp5562
```

`kmod-leds-lp5562` already depends on `kmod-leds-lp55xx-common`

File: package/kernel/linux/modules/leds.mk [hunk 9]
```diff
@@ -261,3 +261,37 @@ define KernelPackage/input-leds/description
 endef
 
 $(eval $(call KernelPackage,input-leds))
+
+
+define KernelPackage/leds-lp55xx-common
```
Comment (ynezz) [inline package/kernel/linux/modules/leds.mk:266]: This kmods needs to be a separate commit, see for example 77a14795dfd5488d62ec5f5cf04b7bd7a8c278ee or generally `git log package/kernel/linux/modules/leds.mk` for prior examples.

**Comment by ynezz** [comment]
> Thank for yours reviews! I squash the commits, add new kernel module package for LED driver and included it on default package for HAZE. If have any mistake please let me know, thank you.

You've probably missed [following part](https://github.com/openwrt/openwrt/pull/14290#issuecomment-1873743943):

> According to the [DTS checklist](https://openwrt.org/submitting-patches#dts_checklist)
> 
> If it's possible try to dedicate some of the LEDs for system status indication in [diag.sh](https://github.com/openwrt/openwrt/blob/master/package/base-files/files/etc/diag.sh) ([example](https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11.dtsi#L7))
> 
> It makes me wonder if it would be possible to define such LED aliases as well?

So either add the LED aliases (preferred) so we get the nice visual LED status feedback during various system phases (booting, upgrading etc.) or you should actually remove the `led_system_*:` labels as there are no users for them.

From https://openwrt.org/docs/guide-developer/device-support-policies#leds

>  The DT label (not the label property) should be created from the node name by adding a led_ prefix, because the DT label will be used in global context. Only add a DT label if you use it for something, otherwise just omit it.

**Review by ynezz — APPROVED** [review]
Thanks! Tested, works, LGTM
